mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h

```diff
@@ -739,6 +740,19 @@ AliasingValueList unknownGetAliasingValues(OpOperand &opOperand);
 /// This is the default implementation of
 /// BufferizableOpInterface::hasTensorSemantics
 bool defaultHasTensorSemantics(Operation *op);
+
+/// This is a helper function used when buffer type is guaranteed to be memref.
+/// It performs two actions: failure state checking and an explicit llvm::cast<>
+/// from the buffer-like type interface to a BaseMemRefType. This allows easier
+/// management of differences in C++ types at the API boundaries. Valid buffer
+/// type is casted to the memref type. Otherwise, the failure state is
+/// propagated i.e. asMemRefType(mlir::failure()) returns mlir::failure().
+FailureOr<BaseMemRefType> asMemRefType(FailureOr<BufferLikeType> bufferType);
```

Comment:
It is weird that you declare this in `detail` namespace, but you use it like public API in other places. My understanding is that the methods within `detail` namespace are not intended to be used widely. With your change, it is used in many other dialects like Linalg/SCF/Arith/etc. Are you going to remove this function? Is it just for the transition state?

Reply:
I see, thanks for the explanation!